### PR TITLE
Try pep 646 on Row

### PR DESCRIPTION
**Pull request by mehdigmira**
draft

### CODE REVIEW

**Comment by zzzeek** [comment]
wow.    OK just so you know that code is generated dynamically by tools/generate_tuple_map_overloads.py and it also applies to sqlalchemy/sql/selectable.py.

I would be surprised if the pep646 code is actually usable, I was never able to get it to do anything useful despite many attempts, but maybe you've figured something out

File: lib/sqlalchemy/engine/row.py [hunk 1]
```diff
@@ -44,10 +45,20 @@
     from .result import RMKeyView
 
 _T = TypeVar("_T", bound=Any)
-_TP = TypeVar("_TP", bound=Tuple[Any, ...])
+_TP = TypeVarTuple("_TP")
+
+if TYPE_CHECKING:
+
+    class _Row(BaseRow, tuple[Unpack[_TP]]):
```
[the remark this answers is not in the document]
Comment (CaselIT) [reply lib/sqlalchemy/engine/row.py:52]: so the main idea is to lie to the type checker making it think we are using a tuple?

**Comment by CaselIT** [comment]
wouldn't we also need to remove this
https://github.com/sqlalchemy/sqlalchemy/blob/4a5831448f1733c80f6b8320f2dbc56627be37ee/lib/sqlalchemy/engine/row.py#L226-L235

File: test/typing/plain_files/orm/session.py [hunk 2]
```diff
@@ -97,6 +97,9 @@ class Address(Base):
         User.id
     ).offset(User.id)
 
+    # EXPECTED_TYPE: Optional[User]
+    reveal_type(sess.query(User).get(1))
```
Comment (eltoder) [inline test/typing/plain_files/orm/session.py:101]: Can use `assert_type` from typing_extensions.

Comment (CaselIT) [reply test/typing/plain_files/orm/session.py:101]: That would need changes to the mypy test fixtures, but maybe it makes sense?

Comment (eltoder) [reply test/typing/plain_files/orm/session.py:101]: Good point. That would be a separate PR.

Comment (CaselIT) [reply test/typing/plain_files/orm/session.py:101]: Indeed. Also myoy would need to print the type in all cases even if the asset is correct

**Comment by mehdigmira** [comment]
closing this until mypy implements pep646 i'll reopen then

**Comment by zzzeek** [comment]
someone has picked this up again in https://github.com/sqlalchemy/sqlalchemy/pull/10634 .  care to take a look ?

**Comment by mehdigmira** [comment]
I've been involved in testing mypy PEP646 changes as they were being developed (for example [here](https://github.com/python/mypy/pull/16131)).
I already have a [MR](https://github.com/mehdigmira/sqlalchemy/pull/2) ongoing that I used as a base for those tests to detect some edge cases. It should be almost ready, I was planning on picking it up again this weekend to polish it.

**Comment by zzzeek** [comment]
OK then we have two people working on the same thing?  What should we do?  can you coordinate on one PR ?  is there a difference in approach?

**Comment by CaselIT** [comment]
seems that the main thing is still making the type-checker thing that Row is a tuple subclass.

On this thing I like the other PR approach better since we avoid the intermediate class https://github.com/sqlalchemy/sqlalchemy/blob/0352cb931f039cc622618d10d1a7cf2b6ba91aef/lib/sqlalchemy/engine/row.py#L42

**Comment by CaselIT** [comment]
> I already have a [MR](https://github.com/mehdigmira/sqlalchemy/pull/2) ongoing that I used as a base for those tests to detect some edge cases. It should be almost ready, I was planning on picking it up again this weekend to polish it.

btw it's fine to open PR directly here, set them in draft / mention that they are so that they are visible also to other people
